### ISSUE TITLE
Stable sorting of DataGrid columns

### ISSUE THREAD

**Issue by laurisvan** [mui/mui-x#10866]
### Search keywords

datagrid stable sort

### Latest version

- [X] I have tested the latest version

### Summary 💡

DataGrid has interesting keyboard modifiers for primary and secondary sorting. However, these are difficult for anybody but heavy users to learn.

Instead, why not use stable sorting when ordering DataGrid contents? For example, see the examples at https://mui.com/x/react-data-grid/sorting/ - when first sorting by "Created on", then "Rating", one would expect to get 5-star rated items on the top, secondarily ordered by "Created on". Instead, the "Created on" seems to get in a seemingly arbitrary order.

This has been a standard behavior for e.g. OS/X and Windows file browsers, Excel etc, so a lot of users have learnt to make use of it.

This probably would also help when inserting items in-between etc, so that the order would be retained.

I am a bit unsure on why the order is not stable, as e.g. JavaScript Array.sort() has been guaranted to be stable as of ECMAScript 2019. What might be the reason DataGrid would not retain its order?

### Motivation 🔦

Providing stable sorting would create a compelling and widely used alternative to the difficult primary and secondary sorting column keyboard modifiers.

### Order ID or Support key 💳 (optional)

69243

**Comment by DanailH** [comment]
@mui/xgrid Can someone help here? I don't think there is a concrete thing that we can do but but rather to understand the decision we took when we did the multi filtering feature.

**Comment by MBilalShafi** [comment]
I am not exactly sure about the motive of this decision, as far as I understand, instead of going with the `primary` and `secondary` only, we chose to allow the users to provide as many criteria as possible, the criteria selected first take higher priority, so in your example, if `Created On` is sorted first, and then `Rating` is sorted, the first preference will be given to the `Created On` which is also illustrated by the numbers assigned to each of the sort.
![image](https://github.com/mui/mui-x/assets/12609561/436bbf13-5500-49e9-a47d-4994d27e7f87)

Now because the `created on` is never repeated, you can see no visual impact on the rating, as it's now a lower priority sort.

Instead, if you try selecting `Rating` first, and then `Created on`, you can see the rows first sorted by rating and then by created date:
![image](https://github.com/mui/mui-x/assets/12609561/ccaa6f4b-067b-4a98-a896-51b68b2f64d7)

Is your proposal to inverse this behavior, i.e. the item sorted on the last taking higher priority? Don't you think it makes more sense to keep it like the current behavior (the users can always swap the order of selection of sorts)?

**Comment by laurisvan** [comment]
@MBilalShafi Let me be clear on what I suggest with the same pictures you have mentioned using good old MS Excel that might be the most widely used "grid editor". 

Plese note that this does not require any special data structures - it only requires that the sorted elements are retained in state, and that when the order is changed, the elements in the state are reordered. Unfortunately I do not know enough of MUI internals that I could point to where the problem might be. But as it fails in your example page, I believe the problem might be specific to the UI component (as mentioned before JavaScript Array.sort is at least claimed to be stable).

1) Initially these people are ordered by name (not relevant in this example)
<img width="354" alt="Screenshot 2023-11-02 at 8 12 51" src="https://github.com/mui/mui-x/assets/1669965/fed7504e-5432-4fb3-9d99-cf61128d91bc">

2) When selecting Sort By "Created on" (desc), the items get ordered from most recent to newest. 
<img width="697" alt="Screenshot 2023-11-02 at 8 13 08" src="https://github.com/mui/mui-x/assets/1669965/af30152d-0dc4-4836-820a-ffe52e372621">
<img width="354" alt="Screenshot 2023-11-02 at 8 13 19" src="https://github.com/mui/mui-x/assets/1669965/552a3a56-227b-4d83-9df1-10c545135e54">

3) When selecting Sort By "Rating", the items get ordered by rating. 

Current & expected outcome: *Please pay close attention to "Created on" - they remain ordered by date within the rating groups (the most recent 5-start item appears first, e.g. 24.10. appears before 12.10.). This is not a coincidence, but stable sorting.
<img width="354" alt="Screenshot 2023-11-02 at 8 33 23" src="https://github.com/mui/mui-x/assets/1669965/630c6c94-6d6d-4d41-875b-2d0cad5bd348">

Now, when repeating the same pattern with the MUI Grid example:
1) The initial order is whatever it might be (in this case Name DESC)

<img width="813" alt="Screenshot 2023-11-02 at 8 14 42" src="https://github.com/mui/mui-x/assets/1669965/59c63390-0bad-47d5-bf7a-818162698bb4">

2) I sort by "Created on" DESC

<img width="813" alt="Screenshot 2023-11-02 at 8 19 03" src="https://github.com/mui/mui-x/assets/1669965/0a32364c-ac0f-4d09-8cd5-2964ed0f28f7">

3) I Sort by "Rating" DESC


<img width="812" alt="Screenshot 2023-11-02 at 8 15 29" src="https://github.com/mui/mui-x/assets/1669965/7b35e534-1057-49f9-8659-4d2ce89c46b7">

When scrolling down, I see':

<img width="816" alt="Screenshot 2023-11-02 at 8 19 21" src="https://github.com/mui/mui-x/assets/1669965/4d8897de-4074-46f1-8f24-de9daadb6420">

Current outcome: The previous sort order is not retained, at least fully - the order of "Created on" is messed up within the same 5-star rating groups
Expected outcome: The previous sort order would be retained

**Comment by MBilalShafi** [comment]
Ah I see, thanks for the detailed example, now I understand your request.

For the grid, you can achieve the same by multiple sort criterion as I mentioned in my first comment, the grid doesn't retain the previous sort for the column from which the sort has been removed, that's a design choice we took while implementing sorting and I think it makes sense too because more freedom is available to users to define any sort priority/order using multi-sorting. (For a benchmark, AgGrid also behaves in a similar fashion)

@mui/xgrid What do you think?

**Comment by cherniavskii** [comment]
Would it make sense to add an option for changing the default sorting behavior to multi-sorting?

**Comment by laurisvan** [comment]
@MBilalShafi Just in order to understand the situation a wee bit better - why would the grid would need to retain the previous sort, if stable sorting is all about retaining the **data** across the sorting operations - not any sorting state?

Anyway, I am not personally that picky on whether it is done technically as stable sort (it just happens to be what I have done everywhere else), as long as the interaction model would not require users to learn keyboard modifiers. Our user group varies from technical experts to absolute tech illiterates, and I doubt the latter even knows which keyboard modifier to use. :)

I remember there was at some point a possilibility to sort ASC/DESC/OFF. I wonder if we could figure out a mouse operated model with these, e.g. if I first sort by one column, and then sort by another column (without clicking a modifier key), effectively still retaining the first column - at least until it is explicitly turned off?

**Comment by MBilalShafi** [comment]
> why would the grid would need to retain the previous sort, if stable sorting is all about retaining the data across the sorting operations - not any sorting state

You are correct, when I mentioned "the grid doesn't retain the previous sort", technically I meant the same, the wording might be confusing. Tbh, I feel your request makes sense.

We discussed and decided to consider the stable sort as a `Pro` option alongside multi-sorting. Please keep checking this issue for further updates.

**Comment by laurisvan** [comment]
@MBilalShafi Thanks for your information. As a happy user of Pro version, I am glad to see this one realised one day.
